### PR TITLE
Migrate in all devteam's DMs

### PR DESCRIPTION
**Pull request by martenson**
part of https://github.com/galaxyproject/tools-devteam/issues/406
xref: https://github.com/galaxyproject/tools-iuc/pull/1060

this is an effort to address part of https://github.com/galaxyproject/tools-devteam/issues/416

### CODE REVIEW

**Comment by bgruening** [comment]
We have no way to test data managers yet, and the tests are failing because of pep8 warnings. @nsoranzo can we exclude dm's from pep8 testing. I don't feel good to change so much without testing in place.

**Comment by nsoranzo** [comment]
@bgruening I started linting them, it's straightforward, the only issue is that they don't specify a Python requirement (which is the case also for other tools in this and other tool repos), so I think they may be executed under miniconda3's Python3 when using 17.01.
We have 3 options:
- support both Python 2/3, which sometimes requires adding a `six` requirement
- support only Python 2 by adding the corresponding requirement (less code changes, but not future-proof)
- migrate them to Python3 and add the corresponding requirement.

Ping @galaxyproject/iuc for opinions.

**Comment by mvdbeek** [comment]
Given that we primarily support conda in the IUC  and the ease with which we can install requirements I think there's no reason the keep python2 around IMO, so I'm voting to port the code to python3.

**Comment by bgruening** [comment]
I'm also for porting, but I really scared about the missing test coverage ...

**Comment by nsoranzo** [comment]
I'm not sure it makes sense to adopt `data_manager_hisat_index_builder`, this should probably be deprecated in favour of `hisat2` and `data_manager_hisat2_index_builder`.

**Comment by nsoranzo** [comment]
Similarly, `data_manager_bwa_index_builder` should be deprecated in favour of `data_manager_bwa_mem_index_builder`.

**Comment by nsoranzo** [comment]
And we may also drop `data_manager_fetch_genome_all_fasta` in favour of the more complete `data_manager_fetch_genome_dbkeys_all_fasta`. Opionions @galaxyproject/iuc?

**Comment by mvdbeek** [comment]
Apart from the compressed portion of `data_manager_fetch_genome_dbkeys_all_fasta` being broken right now (I'll fix this ASAP) I'd be in favor of that.

**Comment by nsoranzo** [comment]
I've pushed a big cleanup which should fix the PEP8 errors for all data managers except the ones which I think we should drop (see above), adding also Python3 compatibility.

File: data_managers/data_manager_fetch_genome_all_fasta/data_manager/data_manager_fetch_genome_all_fasta.py [hunk 1]
```diff
@@ -84,13 +96,14 @@ def _move_and_index_fasta_for_sorting( fasta_filename ):
             line = line.split( None, 1 )[0][1:]
             fasta_offsets[ line ] = offset
     unsorted_fh.close()
-    current_order = map( lambda x: x[1], sorted( map( lambda x: ( x[1], x[0] ), fasta_offsets.items() ) ) )
+    current_order = [_[1] for _ in sorted( ( _[1], _[0] ) for _ in fasta_offsets.items() )]
```
Comment (bgruening) [inline data_managers/data_manager_fetch_genome_all_fasta/data_manager/data_manager_fetch_genome_all_fasta.py]: Nice one! :)

File: data_managers/data_manager_fetch_genome_all_fasta/data_manager/data_manager_fetch_genome_all_fasta.xml [hunk 2]
```diff
@@ -1,67 +1,60 @@
 <tool id="data_manager_fetch_genome_all_fasta" name="Reference Genome" version="0.0.1" tool_type="manage_data">
     <description>fetching</description>
-    <command interpreter="python">data_manager_fetch_genome_all_fasta.py "${out_file}" --dbkey_description ${ dbkey.get_display_text() }</command>
+    <command detect_errors="exit_code">python '$__tool_directory__/data_manager_fetch_genome_all_fasta.py' '${out_file}' --dbkey_description '${dbkey.get_display_text()}'</command>
```
Comment (bgruening) [inline data_managers/data_manager_fetch_genome_all_fasta/data_manager/data_manager_fetch_genome_all_fasta.xml]: Can we break this into multiple lines.

File: data_managers/data_manager_gatk_picard_index_builder/data_manager/data_manager_gatk_picard_index_builder.xml [hunk 3]
```diff
@@ -4,30 +4,28 @@
         <requirement type="package" version="0.1.18">samtools</requirement>
         <requirement type="package" version="1.56.0">picard</requirement>
```
Comment (bgruening) [inline data_managers/data_manager_gatk_picard_index_builder/data_manager/data_manager_gatk_picard_index_builder.xml]: picard and samtools update?

File: data_managers/data_manager_rsync_g2/data_manager/data_manager_rsync.xml [hunk 4]
```diff
@@ -1,23 +1,15 @@
 <tool id="data_manager_rsync_g2" name="Rsync with g2" version="0.0.1" tool_type="manage_data">
     <options sanitize="False" />
     <description>fetching</description>
-    <command interpreter="python">data_manager_rsync.py "${out_file}"</command>
-    <stdio>
-        <exit_code range="1:" err_level="fatal" />
-        <exit_code range=":-1" err_level="fatal" />
-    </stdio>
+    <command detect_errors="exit_code">python '$__tool_directory__/data_manager_rsync.py' '${out_file}'</command>
```
Comment (bgruening) [inline data_managers/data_manager_rsync_g2/data_manager/data_manager_rsync.xml]: multiple lines?

File: data_managers/data_manager_rsync_g2/data_manager/data_manager_rsync.xml [hunk 5]
```diff
@@ -1,23 +1,15 @@
...
-        <param name="data_table_names" type="select" display="checkboxes" multiple="True" optional="True" 
-        label="Choose Desired Data Tables" dynamic_options="galaxy_code_get_available_data_tables( __trans__ )"
-        refresh_on_change="dbkey"/>
-        
-        
-        <param name="data_table_entries" type="select" display="checkboxes" multiple="True" optional="False" 
-        label="Choose Desired Data Tables Entries" dynamic_options="galaxy_code_get_available_data_tables_entries( __trans__, dbkey, data_table_names )"
-        refresh_on_change="dbkey"/>
+        <param name="data_table_names" type="select" display="checkboxes" multiple="True" optional="True" refresh_on_change="dbkey"
+            label="Choose Desired Data Tables" dynamic_options="galaxy_code_get_available_data_tables( __trans__ )" />
```
Comment (bgruening) [inline data_managers/data_manager_rsync_g2/data_manager/data_manager_rsync.xml]: We are adding dynamic_options here into an IUC repo ;)

File: data_managers/data_manager_sam_fasta_index_builder/data_manager/data_manager_sam_fasta_index_builder.xml [hunk 6]
```diff
@@ -3,23 +3,21 @@
     <requirements>
         <requirement type="package" version="0.1.19">samtools</requirement>
     </requirements>
-    <command interpreter="python">data_manager_sam_fasta_index_builder.py "${out_file}" --fasta_filename "${all_fasta_source.fields.path}" --fasta_dbkey "${all_fasta_source.fields.dbkey}" --fasta_description "${all_fasta_source.fields.name}" --data_table_name "fasta_indexes"</command>
+    <command detect_errors="exit_code">python '$__tool_directory__/data_manager_sam_fasta_index_builder.py' '${out_file}' --fasta_filename '${all_fasta_source.fields.path}' --fasta_dbkey '${all_fasta_source.fields.dbkey}' --fasta_description '${all_fasta_source.fields.name}' --data_table_name fasta_indexes</command>
```
Comment (bgruening) [inline data_managers/data_manager_sam_fasta_index_builder/data_manager/data_manager_sam_fasta_index_builder.xml]: multiple lines

File: data_managers/data_manager_twobit_builder/data_manager/twobit_builder.xml [hunk 7]
```diff
@@ -3,14 +3,13 @@
         <requirement type="package" version="324">ucsc-fatotwobit</requirement>
     </requirements>
     <description>builder</description>
-    <command interpreter="python">twobit_builder.py "${out_file}" --fasta_filename "${all_fasta_source.fields.path}" --fasta_dbkey "${all_fasta_source.fields.dbkey}" --fasta_description "${all_fasta_source.fields.name}"</command>
+    <command detect_errors="exit_code">python '$__tool_directory__/twobit_builder.py' '${out_file}' --fasta_filename '${all_fasta_source.fields.path}' --fasta_dbkey '${all_fasta_source.fields.dbkey}' --fasta_description '${all_fasta_source.fields.name}'</command>
```
Comment (bgruening) [inline data_managers/data_manager_twobit_builder/data_manager/twobit_builder.xml]: multiple lines

**Comment by mvdbeek** [comment]
160dd11 should fix the problem @bgruening reported in https://github.com/galaxyproject/tools-devteam/pull/458 and also introduces some doctests at least for the download methods.

**Comment by bgruening** [comment]
Pretty cool, thanks for fixing! Any explanation why this broke? Did UCSC changed something?

**Comment by mvdbeek** [comment]
tar.gz/tar.bz downloads had been broken since I added this functionality (How did I not notice that? I even added test-data to test it ...). I think UCSC previously used zips, but I'm not sure about that.

File: data_managers/data_manager_gatk_picard_index_builder/data_manager/data_manager_gatk_picard_index_builder.xml (hunk 3 above)
Comment (nsoranzo) [inline data_managers/data_manager_gatk_picard_index_builder/data_manager/data_manager_gatk_picard_index_builder.xml]: I am a bit scared about that, that's why I haven't removed the `tool_dependencies.xml`. If you want to take a stab at it, feel free to add commits to this branch.

**Comment by bgruening** [comment]
@martenson can you grant IUC permissions on all these repositories.
@galaxyproject/iuc I excluded data_managers from automatic upload, because we can not test them. But on the other hand no one will test them if they are on master, people will test them on this branch, before merging. So do you think we should enable automatic pushed for DM as well?

**Comment by nsoranzo** [comment]
> So do you think we should enable automatic pushed for DM as well?

Yes, and for packages too (even if there are very few of them being added/updated now).

**Comment by mvdbeek** [comment]
Yes, I think we should push automatically. The master branch should always be the latest and greatest, and I think it would be super annoying to fix bugs if on top of a potential bug there is the uncertainty of whether we may have already fixed that in the master branch.

**Comment by bgruening** [comment]
Thanks @martenson, @nsoranzo, @mvdbeek! I'm ready for merging - anything else I miss?

**Comment by nsoranzo** [comment]
As mentioned above, I've removed the deprecated `data_manager_bwa_index_builder`, `data_manager_fetch_genome_all_fasta` and `data_manager_hisat_index_builder` from this PR, they can safely remain in tools-devteam's `legacy/` directory.

Unfortunately [Travis has issues today](https://www.traviscistatus.com/incidents/f2b5qnpm1y7d), so the build logs seem empty, I've restarted one job.

**Comment by bgruening** [comment]
Looks good. @martenson feel free to merge as soon as you have granted the permissions to IUC.
Thanks.

**Comment by martenson** [comment]
All permissions granted. Should we deprecate the left devteam's DMs?

**Comment by martenson** [comment]
Thanks all for help!

**Pull request merged**